lib/thread.c

```diff
@@ -41,6 +41,14 @@ DEFINE_MTYPE_STATIC(LIB, THREAD_STATS,  "Thread stats")
 #include <mach/mach_time.h>
 #endif
 
+static unsigned char wakebyte = 0x01;
+
+#define ACQUIRE_MASTER(m) \
+  do { \
+    while (pthread_mutex_trylock (&m->mtx)) \
+      write (m->io_pipe[1], &wakebyte, 1); \
+  } while (0);
+
```

Comment:
With the current design this should function correctly, but the design is seriously bad. This will block in write() after filling up the kernel's pipe buffer, i.e 4096 useless syscalls later.

The better design is for `thread_fetch()` to not hold `m->mtx` during the `fd_select` call. There is no reason for it to do so; none of the working data of the `fd_select` call can be modified by another thread. If `thread_fetch` is restructured this way, the pipe-poke can be done (unconditionally) at mutex _release_ time instead. The purpose slightly changes: instead of poking to get the mutex, it's a poke to update the `fd_select` arguments to in-/exclude new items.